python/dgl/transform/module.py

```diff
@@ -0,0 +1,790 @@
+    and return.
+
+    The function transforms the coordinates/features of a point set
+    into a directed homogeneous graph. The coordinates of the point
+    set is specified as a matrix whose rows correspond to points and
+    columns correspond to coordinate/feature dimensions.
+
+    The nodes of the returned graph correspond to the points, where the predecessors
+    of each point are its k-nearest neighbors measured by the chosen distance.
+
+    Parameters
+    ----------
+    ndata_name : str
+        The ndata name to store the node features for KNN computation.
+    k : int
+        The number of nearest neighbors per node.
+    algorithm : str, optional
+        Algorithm used to compute the k-nearest neighbors.
+
+        * 'bruteforce-blas' will first compute the distance matrix
+          using BLAS matrix multiplication operation provided by
+          backend frameworks. Then use topk algorithm to get
+          k-nearest neighbors. This method is fast when the point
+          set is small but has :math:`O(N^2)` memory complexity where
+          :math:`N` is the number of points.
+
+        * 'bruteforce' will compute distances pair by pair and
+          directly select the k-nearest neighbors during distance
+          computation. This method is slower than 'bruteforce-blas'
+          but has less memory overhead (i.e., :math:`O(Nk)` where :math:`N`
+          is the number of points, :math:`k` is the number of nearest
+          neighbors per node) since we do not need to store all distances.
+
+        * 'bruteforce-sharemem' (CUDA only) is similar to 'bruteforce'
+          but use shared memory in CUDA devices for buffer. This method is
+          faster than 'bruteforce' when the dimension of input points
+          is not large. This method is only available on CUDA device.
+
+        * 'kd-tree' will use the kd-tree algorithm (CPU only).
+          This method is suitable for low-dimensional data (e.g. 3D
+          point clouds)
+
+        * 'nn-descent' is an approximate approach from paper
+          `Efficient k-nearest neighbor graph construction for generic similarity
+          measures <https://www.cs.princeton.edu/cass/papers/www11.pdf>`_. This method
+          will search for nearest neighbor candidates in "neighbors' neighbors".
+
+        (default: 'bruteforce-blas')
+    dist : str, optional
+        The distance metric used to compute distance between points. It can be the following
+        metrics:
+        * 'euclidean': Use Euclidean distance (L2 norm) :math:`\sqrt{\sum_{i} (x_{i} - y_{i})^{2}}`.
+        * 'cosine': Use cosine distance.
+        (default: 'euclidean')
+
+    Example
+    -------
+
+    The following examples use PyTorch backend.
+
+    >>> import dgl
+    >>> import torch
+    >>> from dgl import KNNGraph
+
+    >>> g = dgl.rand_graph(5, 20)
+    >>> g.ndata['h'] = torch.randn(g.num_nodes(), 3)
+    >>> transform = KNNGraph(ndata_name='h', k=3)
```

Comment:
This is a little bit strange. Shouldn't it transform a tensor to a graph?

Reply:
In that case, it cannot be used in `Compose` as well.

Reply:
True. Then why should we have this? How people use it in real cases?

Reply:
Removed this module.